larpandoracontent/LArTrackShowerId/SvmPfoCharacterisationAlgorithm.h

```diff
@@ -25,31 +25,62 @@ class SvmPfoCharacterisationAlgorithm : public PfoCharacterisationBaseAlgorithm
      */
     SvmPfoCharacterisationAlgorithm();
 
+    /**
+     *  @brief Destructor
+     */
+    ~SvmPfoCharacterisationAlgorithm();
+class RecoParameters
+{
+public:
+        /**
+         *  @brief  Constructor
+         */
+     RecoParameters();
+
+     unsigned int  m_minPrimaryGoodHits;       ///< the minimum number of primary good Hits
+     unsigned int  m_minHitsForGoodView;       ///< the minimum number of Hits for a good view
+     unsigned int  m_minPrimaryGoodViews;      ///< the minimum number of primary good views
+	 bool          m_foldToPrimaries;          ///< whether to fold all hits to primary pfos and MC particles
+     float         m_minHitSharingFraction;    ///< the minimum Hit sharing fraction
+    };
 protected:
+
+    pandora::StatusCode Run();
     virtual bool IsClearTrack(const pandora::ParticleFlowObject *const pPfo) const;
     virtual bool IsClearTrack(const pandora::Cluster *const pCluster) const;
+    void FillMCToRecoHitsMap(const pandora::MCParticleList *pMCParticleList, const pandora::CaloHitList *pCaloHitList, LArMCParticleHelper::MCContributionMap &mcToRecoHitsMap) const;
+    void SelectParticlesByHitCount(const pandora::MCParticleVector &candidateTargets, const LArMCParticleHelper::MCContributionMap &mcToTrueHitListMap, const LArMCParticleHelper::MCRelationMap &mcToTargetMCMap, LArMCParticleHelper::MCContributionMap &selectedMCParticlesToHitsMap) const;
+
+    void SelectGoodCaloHits(const pandora::CaloHitList *const pSelectedCaloHitList, const LArMCParticleHelper::MCRelationMap &mcToTargetMCMap, pandora::CaloHitList &selectedGoodCaloHitList) const;
+    void GetMCToSelfMap(const pandora::MCParticleList *const pMCParticleList, LArMCParticleHelper::MCRelationMap &mcToSelfMap) const;
     pandora::StatusCode ReadSettings(const pandora::TiXmlHandle xmlHandle);
 
     ClusterCharacterisationFeatureTool::FeatureToolVector   m_featureToolVector;         ///< The feature tool map
     PfoCharacterisationFeatureTool::FeatureToolVector       m_featureToolVectorThreeD;   ///< The feature tool map for 3D info
     PfoCharacterisationFeatureTool::FeatureToolVector       m_featureToolVectorNoChargeInfo; ///< The feature tool map for missing W view
 
-    SupportVectorMachine    m_supportVectorMachine;              ///< The support vector machine
-    SupportVectorMachine    m_supportVectorMachineNoChargeInfo;  ///< The support vector machine for missing W view
+    AdaBoostDecisionTree    m_adaBoostDecisionTree;              ///< The Boosted Decision Tree
+    AdaBoostDecisionTree    m_adaBoostDecisionTreeNoChargeInfo;  ///< The Boosted Decision Tree for missing W view
```

Comment:
As discussed with @AndyChappell, we want to be able to support both a Svm and a Bdt here, so we need to template on the Mva type, as previously done e.g. [here](https://github.com/PandoraPFA/LArContent/commit/56ccc12ac313aea355a5f8fd1bd80f34496daf01). (We might want to support multiple XML configuration keys to avoid breaking config changes.)